text/0000-code-literals.md

```diff
@@ -0,0 +1,517 @@
+in a single canonical way, while allowing the indentation of the string
+as a whole to be changed freely.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+An `h` modifier may be added to the prefix of the following string
+literal types:
+
+- String literals `h"`
+- Raw string literals `hr#"`
+- Byte string literals `hb"`
+- Raw byte string literals `hbr#"`
+
+The `h` modifier will appear before all characters in the prefix.
+
+The value of a string literal with the `h` modifier will be determined
+using the following steps:
+
+1.  Measure the whitespace indenting the closing quote. If a
+    non-whitespace character (other than a single `-`) exists before
+    the closing quote on the same line, then issue a compiler error.
+2.  Take the lines *between* (but not including) the opening and
+    closing quotes exactly as written.
+3.  Remove exactly the measured whitespace from each non-empty line.
+    If this cannot be done, then issue a compiler error. The
+    whitespace must match down to the exact character sequence.
+4.  If a `-` character was present immediately prior to the closing
+    quote, then remove the final newline.
+5.  Interpret any escape sequences and apply any pre-processing as
+    usual for the string literal type without an `h` modifier.
+    For example, newlines in the file are always treated as `\n`
+    even if the file is encoded with `\r\n` newlines.
+
+Here are some edge case examples:
+
+```rust
+    // Empty string with language hint
+    assert_eq!(h"foo
+        ", "");
+
+    // Newline
+    assert_eq!(h"
+
+        ", "\n");
+
+    // No terminating newline
+    assert_eq!(h"
+        bar
+        -", "bar");
+
+    // Terminating newline
+    assert_eq!(h"
+        bar
+        ", "bar\n");
+
+    // Preserved indent
+    assert_eq!(hr#"
+    if a:
+        print(42)
+    "#, "if a:\n    print(42)\n");
+
+    // Relative indent
+    assert_eq!(hr#"
+            if a:
+                print(42)
+            "#, "if a:\n    print(42)\n");
+
+    // Relative to closing quote
+    assert_eq!(hr#"
+
+
+            if a:
+                print(42)
+    "#, "\n\n        if a:\n            print(42)\n");
+
+    // Interactions with escaping rules
+    assert_eq!(h"
+        \"\
+            foo\n
+            bar
+    \t
+    ", "    \"foo\n\n        bar\n\t\n");
+```
+
+Any text between the opening quote and the first newline is
+preserved within the AST, but is otherwise unused.
+
+This is a backwards compatible change for editions 2021 onwards, since
+edition 2021 reserved prefixes for this kind of feature:
+https://doc.rust-lang.org/reference/tokens.html#reserved-prefixes.
+
+Editions prior to 2021 will not benefit from this feature.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback is increased complexity of the language:
+
+1.  It adds a four new types of string literals given all
+    the combinations.
```

Comment:
Teaching.

String literals are used in pattern matching. It will be very annoying to explain why a metadata tag that can be written as part of the literal and lives inside what *appears to be* the string's delimiter tokens does or **does not** participate in pattern matching. I would prefer the question simply not arise.

Reply:
Sure, we could add more complex rule for `language_tag`, for example first line with a tag must end with `#`.
```rust
let h#"sql#
    SELECT
    "# = &maybe_sql_expr[0..2] else { return; };
```

Reply:
You have not actually clarified **anything** as long as the tag is inside the quotation marks.

Reply:
> But I do not want to explain why either of these may or may not work:

As proposed in the RFC, both of those would match as the language hint is not part of the value. I don't think this case is really any different from eg.

```rust
let "\t" = &maybe_tab_expr[0..0] else {
    return;
};

let "	" = &maybe_tab_expr[0..0] else {
    return;
};
```

Or:
```rust
enum Foo {
    Bar,
    Baz,
}

use Foo::Bar as Bat;


fn main() {
    match Foo::Bar {
        self::Bat => println!("Bat"),
        Foo::Baz => println!("Baz")
    }
}
```

Ultimately, you can't expect pattern matching to be syntactic - it's fundamentally about the value.

Reply:
Yes yes, and `2_5_5` also is matched by `255` and `0xFF`, as those names alias, and if you introduce a specific alias for something, shockingly, it matches. And introducing redundant aliases without consideration for the potential harms to understanding is what I am objecting.

However, your comments have made apparent to me that you fundamentally do not actually believe this increases language complexity, as you don't think it makes it harder to parse or understand the source code, so I *also* object to the text written here. If mere quantitative increase in ways to express something does not count as an increase in complexity, then this entry is a lie and there is no drawback.

```suggestion
None.
```

Reply:
> However, your comments have made apparent to me that you fundamentally do not actually believe this increases language complexity, as you don't think it makes it harder to parse or understand the source code

I do think it increases language complexity, but only in the sense that the language now has N+1 features rather than N. Where I disagree with you is the idea that there is a qualitative rather than quantitative difference in complexity in comparison to existing string literals.

My hope is that even this incremental increase in complexity could be later reduced: given that the feature is designed to allow represent every possible string, I think there's a world where a future edition simply makes all multiline literals behave like the literals proposed here. I think it would be appropriate to propose this more drastic change if we later find that the use of code string literals naturally replaces the use of multiline string / raw string literals due to people preferring an "indentation relative form", and if no unforeseen drawbacks are encountered.